cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,440 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Adapted from Asylo
+
+#pragma once
+
+#include <string>
+#include <utility>
+
+#include "arrow/status.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
+
+namespace internal {
+void DieWithMessage(const std::string& msg);
+}  // namespace internal
+
+struct ARROW_EXPORT ErrorOrConstants {
+  static const char kValueMoveConstructorMsg[];
+  static const char kValueMoveAssignmentMsg[];
+  static const char kValueOrDieMovedMsg[];
+  static const char kStatusMoveConstructorMsg[];
+  static const char kStatusMoveAssignmentMsg[];
+};
+
+// A class for representing either a usable value, or an error.
+///
+/// A ErrorOr object either contains a value of type `T` or a Status object
+/// explaining why such a value is not present. The type `T` must be
+/// copy-constructible and/or move-constructible.
+///
+/// The state of a ErrorOr object may be determined by calling ok() or
+/// status(). The ok() method returns true if the object contains a valid value.
+/// The status() method returns the internal Status object. A ErrorOr object
+/// that contains a valid value will return an OK Status for a call to status().
+///
+/// A value of type `T` may be extracted from a ErrorOr object through a call
+/// to ValueOrDie(). This function should only be called if a call to ok()
+/// returns true. Sample usage:
+///
+/// ```
+///   arrow::ErrorOr<Foo> result = CalculateFoo();
+///   if (result.ok()) {
+///     Foo foo = result.ValueOrDie();
+///     foo.DoSomethingCool();
+///   } else {
+///     ARROW_LOG(ERROR) << result.status();
+///  }
+/// ```
+///
+/// If `T` is a move-only type, like `std::unique_ptr<>`, then the value should
+/// only be extracted after invoking `std::move()` on the ErrorOr object.
+/// Sample usage:
+///
+/// ```
+///   arrow::ErrorOr<std::unique_ptr<Foo>> result = CalculateFoo();
+///   if (result.ok()) {
+///     std::unique_ptr<Foo> foo = std::move(result).ValueOrDie();
+///     foo->DoSomethingCool();
+///   } else {
+///     ARROW_LOG(ERROR) << result.status();
+///   }
+/// ```
+///
+/// ErrorOr is provided for the convenience of implementing functions that
+/// return some value but may fail during execution. For instance, consider a
+/// function with the following signature:
+///
+/// ```
+///   arrow::Status CalculateFoo(int *output);
+/// ```
+///
+/// This function may instead be written as:
+///
+/// ```
+///   arrow::ErrorOr<int> CalculateFoo();
+/// ```
+template <class T>
+class ARROW_EXPORT ErrorOr {
```

Comment:
My 2 cents, but I find this name ugly. `Result<T>` reads much better.

Reply:
My issue with Result is it isn't super descriptive but if that becomes a blocker we can change it

Reply:
`Result` is used by rust, ocaml, f#. Haskell uses `Either`. Scala uses `Try`.

Reply:
`StatusOr` might be ok too

Reply:
Hmm, out of all of these if I had to swith, I think I would use "Try" since it is the shortest.  Can we roll a die somehow to choose?